File: Sample/SampleOC/LandscapeViewController.m
//
//  LandscapeViewController.m
//  Sample
//
//  Created by KK on 2026/6/28.
//

#import "LandscapeViewController.h"

@import OrientationTransitionKit;

@interface LandscapeViewController () <OTKTransitionToContextProvider>
@property (nonatomic, strong) UIView *playerContainerView;
@property (nonatomic, strong) UIButton *closeButton;
@property (nonatomic, strong) UIButton *userInfoButton;
@property (nonatomic, assign) BOOL shouldHideHomeIndicator;
@end

@implementation LandscapeViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.view.backgroundColor = UIColor.blackColor;

    _playerContainerView = [UIView new];
    _playerContainerView.backgroundColor = UIColor.blackColor;
    _playerContainerView.translatesAutoresizingMaskIntoConstraints = NO;

    _closeButton = [UIButton buttonWithType:UIButtonTypeCustom];
    _closeButton.backgroundColor = [UIColor colorWithWhite:0 alpha:0.55];
    _closeButton.layer.cornerRadius = 6;
    _closeButton.clipsToBounds = YES;
    [_closeButton setTitle:@"退出全屏" forState:UIControlStateNormal];
    [_closeButton setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
    _closeButton.titleLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightMedium];
    _closeButton.translatesAutoresizingMaskIntoConstraints = NO;
    [_closeButton addTarget:self action:@selector(closeButtonTapped) forControlEvents:UIControlEventTouchUpInside];

    _userInfoButton = [UIButton buttonWithType:UIButtonTypeCustom];
    _userInfoButton.backgroundColor = [UIColor colorWithWhite:0 alpha:0.55];
    _userInfoButton.layer.cornerRadius = 6;
    _userInfoButton.clipsToBounds = YES;
    [_userInfoButton setTitle:@"用户主页" forState:UIControlStateNormal];
    [_userInfoButton setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
    _userInfoButton.titleLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightMedium];
    _userInfoButton.translatesAutoresizingMaskIntoConstraints = NO;
    [_userInfoButton addTarget:self action:@selector(userInfoButtonTapped) forControlEvents:UIControlEventTouchUpInside];

    [self.view addSubview:_playerContainerView];
    [self.view addSubview:_closeButton];
    [self.view addSubview:_userInfoButton];

    [NSLayoutConstraint activateConstraints:@[
        [_playerContainerView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [_playerContainerView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [_playerContainerView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor],
        [_playerContainerView.heightAnchor constraintEqualToAnchor:self.view.heightAnchor],

        [_closeButton.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor constant:20],
        [_closeButton.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:20],
        [_closeButton.widthAnchor constraintEqualToConstant:96],
        [_closeButton.heightAnchor constraintEqualToConstant:44],

        [_userInfoButton.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor constant:-20],
        [_userInfoButton.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:20],
        [_userInfoButton.widthAnchor constraintEqualToConstant:96],
        [_userInfoButton.heightAnchor constraintEqualToConstant:44],
    ]];
}

- (BOOL)shouldAutorotate {
    return YES;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    return UIInterfaceOrientationMaskLandscapeRight;
}

- (UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
    return UIInterfaceOrientationLandscapeRight;
}

- (BOOL)prefersHomeIndicatorAutoHidden {
    return self.shouldHideHomeIndicator;
}

- (void)closeButtonTapped {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)userInfoButtonTapped {
    if (self.didTapUserInfoHandler != nil) {
        self.didTapUserInfoHandler(self);
    }
}

- (void)setHomeIndicatorHidden:(BOOL)hidden {
    self.shouldHideHomeIndicator = hidden;
    [self setNeedsUpdateOfHomeIndicatorAutoHidden];
}

- (UIViewController *)transitionToContextProviderViewController:(id<OTKTransitionToContextProvider>)contextProvider {
    return self;
}

- (CGRect)transitionToContextProviderTransitionFrame:(id<OTKTransitionToContextProvider>)contextProvider in:(UIView *)containerView {
    return [containerView convertRect:self.playerContainerView.bounds fromView:self.playerContainerView];
}

- (void)transitionToContextProviderPrepareTransitionView:(id<OTKTransitionToContextProvider>)contextProvider transitionView:(UIView *)transitionView {
    [self moveTransitionContentToContainerView:transitionView];
}

- (void)transitionToContextProviderFinishTransitionView:(id<OTKTransitionToContextProvider>)contextProvider {
    [self moveTransitionContentToContainerView:self.playerContainerView];
}

- (void)transitionToContextProviderTransitionWillEnter:(id<OTKTransitionToContextProvider>)contextProvider from:(id<OTKTransitionFromContextProvider>)fromContextProvider {
    [self setHomeIndicatorHidden:YES];
}

- (void)transitionToContextProviderTransitionDidEnter:(id<OTKTransitionToContextProvider>)contextProvider from:(id<OTKTransitionFromContextProvider>)fromContextProvider {
    [self setHomeIndicatorHidden:YES];
}

- (void)transitionToContextProviderTransitionWillExit:(id<OTKTransitionToContextProvider>)contextProvider from:(id<OTKTransitionFromContextProvider>)fromContextProvider {
    [self setHomeIndicatorHidden:YES];
}

- (void)transitionToContextProviderTransitionDidExit:(id<OTKTransitionToContextProvider>)contextProvider from:(id<OTKTransitionFromContextProvider>)fromContextProvider {
    [self setHomeIndicatorHidden:NO];
}

- (void)moveTransitionContentToContainerView:(UIView *)containerView {
    UIView *transitionContentView = self.transitionContentView;
    if (transitionContentView == nil) {
        return;
    }

    [transitionContentView removeFromSuperview];
    transitionContentView.transform = CGAffineTransformIdentity;
    transitionContentView.frame = containerView.bounds;
    transitionContentView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    [containerView addSubview:transitionContentView];
}

@end

File: Sample/Sample/LandscapeViewController.swift
//
//  LandscapeViewController.swift
//  Sample
//
//  Created by KK on 2026/6/28.
//

import OrientationTransitionKit
import UIKit

class LandscapeViewController: BaseViewController {
    let playerContainerView = UIView()
    private let closeButton = UIButton(type: .custom)
    private let userInfoButton = UIButton(type: .custom)

    var didTapUserInfoHandler: ((LandscapeViewController) -> Void)?
    weak var transitionContentView: UIView?
    private var shouldHideHomeIndicator = false

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = .black

        playerContainerView.backgroundColor = .black
        playerContainerView.translatesAutoresizingMaskIntoConstraints = false

        closeButton.backgroundColor = UIColor(white: 0, alpha: 0.55)
        closeButton.layer.cornerRadius = 6
        closeButton.clipsToBounds = true
        closeButton.setTitle("退出全屏", for: .normal)
        closeButton.setTitleColor(.white, for: .normal)
        closeButton.titleLabel?.font = .systemFont(ofSize: 16, weight: .medium)
        closeButton.translatesAutoresizingMaskIntoConstraints = false
        closeButton.addTarget(self, action: #selector(closeButtonTapped), for: .touchUpInside)

        userInfoButton.backgroundColor = UIColor(white: 0, alpha: 0.55)
        userInfoButton.layer.cornerRadius = 6
        userInfoButton.clipsToBounds = true
        userInfoButton.setTitle("用户主页", for: .normal)
        userInfoButton.setTitleColor(.white, for: .normal)
        userInfoButton.titleLabel?.font = .systemFont(ofSize: 16, weight: .medium)
        userInfoButton.translatesAutoresizingMaskIntoConstraints = false
        userInfoButton.addTarget(self, action: #selector(userInfoButtonTapped), for: .touchUpInside)

        view.addSubview(playerContainerView)
        view.addSubview(closeButton)
        view.addSubview(userInfoButton)

        NSLayoutConstraint.activate([
            playerContainerView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            playerContainerView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            playerContainerView.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            playerContainerView.heightAnchor.constraint(equalTo: view.heightAnchor),

            closeButton.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 20),
            closeButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
            closeButton.widthAnchor.constraint(equalToConstant: 96),
            closeButton.heightAnchor.constraint(equalToConstant: 44),

            userInfoButton.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -20),
            userInfoButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
            userInfoButton.widthAnchor.constraint(equalToConstant: 96),
            userInfoButton.heightAnchor.constraint(equalToConstant: 44),
        ])
    }

    override var shouldAutorotate: Bool {
        true
    }

    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        .landscapeRight
    }

    override var preferredInterfaceOrientationForPresentation: UIInterfaceOrientation {
        .landscapeRight
    }

    override var prefersHomeIndicatorAutoHidden: Bool {
        shouldHideHomeIndicator
    }

    @objc private func closeButtonTapped() {
        dismiss(animated: true)
    }

    @objc private func userInfoButtonTapped() {
        didTapUserInfoHandler?(self)
    }

    private func setHomeIndicatorHidden(_ isHidden: Bool) {
        shouldHideHomeIndicator = isHidden
        setNeedsUpdateOfHomeIndicatorAutoHidden()
    }
}

extension LandscapeViewController: TransitionToContextProvider {
    func transitionToContextProviderViewController(_ contextProvider: TransitionToContextProvider) -> UIViewController {
        self
    }

    func transitionToContextProviderTransitionFrame(_ contextProvider: TransitionToContextProvider, in containerView: UIView) -> CGRect {
        containerView.convert(playerContainerView.bounds, from: playerContainerView)
    }

    func transitionToContextProviderPrepareTransitionView(_ contextProvider: TransitionToContextProvider, transitionView: UIView) {
        moveTransitionContent(to: transitionView)
    }

    func transitionToContextProviderFinishTransitionView(_ contextProvider: TransitionToContextProvider) {
        moveTransitionContent(to: playerContainerView)
    }

    func transitionToContextProviderTransitionWillEnter(_ contextProvider: TransitionToContextProvider, from fromContextProvider: TransitionFromContextProvider) {
        setHomeIndicatorHidden(true)
    }

    func transitionToContextProviderTransitionDidEnter(_ contextProvider: TransitionToContextProvider, from fromContextProvider: TransitionFromContextProvider) {
        setHomeIndicatorHidden(true)
    }

    func transitionToContextProviderTransitionWillExit(_ contextProvider: TransitionToContextProvider, from fromContextProvider: TransitionFromContextProvider) {
        setHomeIndicatorHidden(true)
    }

    func transitionToContextProviderTransitionDidExit(_ contextProvider: TransitionToContextProvider, from fromContextProvider: TransitionFromContextProvider) {
        setHomeIndicatorHidden(false)
    }

    private func moveTransitionContent(to containerView: UIView) {
        guard let transitionContentView else {
            return
        }

        transitionContentView.removeFromSuperview()
        transitionContentView.transform = .identity
        transitionContentView.frame = containerView.bounds
        transitionContentView.translatesAutoresizingMaskIntoConstraints = false
//        transitionContentView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        containerView.addSubview(transitionContentView)
        NSLayoutConstraint.activate([
            transitionContentView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            transitionContentView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
            transitionContentView.topAnchor.constraint(equalTo: containerView.topAnchor),
            transitionContentView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor),
        ])
    }
}
